File: SqlExpressions/Expressions/SqlConstantExpression.cs
namespace SqlExpressions.Expressions
{
    /// <summary>
    /// Expression representing constant value.
    /// </summary>
    /// <seealso cref="SqlExpressions.Expressions.SqlExpression" />
    public class SqlConstantExpression:SqlExpression
    {
        /// <summary>
        /// Gets the return type of the expression.
        /// </summary>
        /// <value>
        /// The type of the return..
        /// </value>
        public override SqlType ReturnType { get; }
        /// <summary>
        /// Gets the type of the expression.
        /// </summary>
        /// <value>
        /// The type of the expression.
        /// </value>
        public override SqlExpressionType ExpressionType => SqlExpressionType.Constant;
        /// <summary>
        /// Accepts the specified visitor.
        /// </summary>
        /// <typeparam name="T">Return type.</typeparam>
        /// <param name="visitor">The visitor.</param>
        /// <returns>Returns processed value from expression.</returns>
        public override T Accept<T>(ISqlVisitor<T> visitor)
        {
            return visitor.Visit(this);
        }
        /// <summary>
        /// Constant value.
        /// </summary>
        /// <value>
        /// The constant.
        /// </value>
        public object Constant { get; private set; }

        /// <summary>
        /// Initializes a new instance of the <see cref="SqlConstantExpression"/> class.
        /// </summary>
        /// <param name="returnType">Type of the constant.</param>
        /// <param name="constant">The constant.</param>
        public SqlConstantExpression(SqlType returnType, object constant)
        {
            ReturnType = returnType;
            Constant = constant;
        }
    }
}

File: Website/ClientApp/expression-tree/SqlConstantExpression.ts
import { SqlExpression, SqlExpressionBuilder } from "./SqlExpression";
import { SqlType } from "./SqlType";
import { SqlExpressionType } from "./SqlExpressionType";
import { TransferObjectId } from "./TransferObjectId";

/*
 * node in expression tree which holds a constant
 */
export interface SqlConstantExpression extends SqlExpression {
    Constant: any;
}

export class SqlContantExpressionBuilder {
    Constant: any;
    protected ReturnType: SqlType;

    setReturnType(value: SqlType): SqlContantExpressionBuilder {
        this.ReturnType = value;
        return this;
    }

    setConstant(value: string): SqlContantExpressionBuilder {
        this.Constant = value;
        return this;
    }

    build(): SqlConstantExpression {
        return {
            ExpressionType: SqlExpressionType.Constant,
            ReturnType: this.ReturnType,
            Constant: this.Constant,
            TransferObjectId: TransferObjectId.SqlConstantExpression
        };
    }
}
